### PR TITLE
[SPARK-32138] Drop Python 2.7, 3.4 and 3.5

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR aims to drop Python 2.7, 3.4 and 3.5.

Roughly speaking, it removes all the widely known Python 2 compatibility workarounds such as `sys.version` comparison, `__future__`. Also, it removes the Python 2 dedicated codes such as `ArrayConstructor` in Spark.

### Why are the changes needed?

 1. Unsupport EOL Python versions
 2. Reduce maintenance overhead and remove a bit of legacy codes and hacks for Python 2.
 3. PyPy2 has a critical bug that causes a flaky test, SPARK-28358 given my testing and investigation. 
 4. Users can use Python type hints with Pandas UDFs without thinking about Python version
 5. Users can leverage one latest cloudpickle, https://github.com/apache/spark/pull/28950. With Python 3.8+ it can also leverage C pickle.

### Does this PR introduce _any_ user-facing change?

Yes, users cannot use Python 2.7, 3.4 and 3.5 in the upcoming Spark version.

### How was this patch tested?

Manually tested and also tested in Jenkins.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
I'm supporting @HyukjinKwon 's idea, but cc @gatorsmile and @marmbrus since they wanted to keep the deprecated features forever if there is no big burden.

**Comment by dongjoon-hyun** [comment]
Also, cc @rxin

**Review by Fokko — COMMENTED** [review]
I'm in favor of dropping Python <3.6 as it would simplify the code. This also allows us to use type hinting to make pyspark easier too use.

I think, we need to update the classifiers as well: https://github.com/apache/spark/blob/3efa5218ebf978143ecb78ded9e766a1af8e83fc/python/setup.py#L214-L226

Also, I would recommend setting [`python_requires`](https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires) to `>=3.6` so people get a notification if they install it using `<=3.5`.

**Comment by holdenk** [comment]
Thanks for working on this! @fokko do we have a rough idea what % of general users are running Python 3.6+?

**Comment by Fokko** [comment]
My pleasure @holdenk 

I ran a query against the public dataset of Google. They have a dataset that contains all the public pypi downloads:
```sql
SELECT 
  EXTRACT(YEAR FROM timestamp) AS year,
  EXTRACT(MONTH FROM timestamp) AS month,
  SAFE.SUBSTR(details.python, 0, 3) AS python_version,
  COUNT(*) AS num_downloads
FROM `the-psf.pypi.downloads*`
WHERE file.project = 'pyspark'
AND SAFE.SUBSTR(details.python, 0, 3) IS NOT NULL
GROUP BY 
  EXTRACT(YEAR FROM timestamp),
  EXTRACT(MONTH FROM timestamp),
  SAFE.SUBSTR(details.python, 0, 3)
```

This gives us the following per month:
image

We can see that the majority uses 3.7 and 3.6. However, there is still a share of 3.5 and 2.7.

If we look at the proportional share of people who'm using a compatible version:

```sql
SELECT 
  EXTRACT(YEAR FROM timestamp) AS year,
  EXTRACT(MONTH FROM timestamp) AS month,
  if(SAFE.SUBSTR(details.python, 0, 3) >= '3.6', 'ok', 'not_ok') as OK,
  COUNT(*) AS num_downloads
FROM `the-psf.pypi.downloads*`
WHERE file.project = 'pyspark'
AND SAFE.SUBSTR(details.python, 0, 3) IS NOT NULL
GROUP BY 
  EXTRACT(YEAR FROM timestamp),
  EXTRACT(MONTH FROM timestamp),
  if(SAFE.SUBSTR(details.python, 0, 3) >= '3.6', 'ok', 'not_ok')
```

Then the majority is ok:
image

The next question would be if Python <3.6 users are on 3.0 or on 2.x. My guess would be the latter, so we're (mostly) safe deprecating the old versions of Python.

**Comment by HyukjinKwon** [comment]
Thanks @Fokko for the investigation here.

**Comment by dongjoon-hyun** [comment]
Wow, nice investigation, @Fokko ! Thanks.

**Comment by dongjoon-hyun** [comment]
@HyukjinKwon . Did you get a chance to get a feedback from @marmbrus and @gatorsmile ?

**Comment by gatorsmile** [comment]
cc @mateiz and @rxin

**Comment by HyukjinKwon** [comment]
Nope, I am waiting for more feedback - I usually share anything up in OSS side. I was just assuming it's good enough to go given @Fokko's investigation at https://github.com/apache/spark/pull/28957#issuecomment-652022449.

**Comment by HyukjinKwon** [comment]
I will send an email to dev list to confirm. I think that's faster.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

File: python/pyspark/sql/tests/test_context.py
Comment (ueshin) [inline python/pyspark/sql/tests/test_context.py:22]: We can remove the comment `# Python 3.4+ only.`?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

File: docs/index.md
Comment (viirya) [inline docs/index.md:47]: Python 2.7+/3.4+ -> Python 3.6+?

File: docs/rdd-programming-guide.md
Comment (viirya) [inline docs/rdd-programming-guide.md:104]: works with Python 3.6+?

File: docs/rdd-programming-guide.md
Comment (HyukjinKwon) [inline docs/rdd-programming-guide.md:277]: BTW, this seems broken in Spark due to how [find-spark-home](https://github.com/apache/spark/blob/master/bin/find-spark-home) works in PySpark, see https://github.com/apache/spark/pull/28256

File: docs/rdd-programming-guide.md
Comment (HyukjinKwon) [inline docs/rdd-programming-guide.md:281]: BTW, Jupyter with PySpark in this way seems broken in Spark due to how [find-spark-home](https://github.com/apache/spark/blob/master/bin/find-spark-home) works in PySpark, see https://github.com/apache/spark/pull/28256

File: python/pyspark/sql/types.py
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:1490]: cc @BryanCutler to double check. Basically, I removed the logic here to don't sort keyword arguments by default, and removed `__from_dict__` related logics too.

With Python 3.6+, it (already) broke one test case here: https://github.com/apache/spark/pull/28957/files#diff-0590ca852e0e565bc489272aee36167fL890 but I guess it's fine.

Comment (HyukjinKwon) [reply python/pyspark/sql/types.py:1490]: cc @zero323 too FYI. Finally we're removing this.

File: python/pyspark/sql/dataframe.py
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py:2338]: So this is also because of https://github.com/apache/spark/pull/28957/files#r449475633

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
cc @BryanCutler, @ueshin, @holdenk, @dongjoon-hyun, @viirya, @srowen, this should be ready for a review.

**Review by srowen — COMMENTED** [review]
Looks OK pending tests

File: dev/lint-python
Comment (viirya) [inline dev/lint-python:210]: Comment it by purpose? Just remove it?

File: dev/run-tests.py
Comment (viirya) [inline dev/run-tests.py:652]: Is this related to dropping Python 2.7 support?

File: python/run-tests.py
Comment (viirya) [inline python/run-tests.py:164]: pypy?

**Comment by viirya** [comment]
Looks good overall.

File: dev/run-tests.py
Comment (HyukjinKwon) [reply dev/run-tests.py:652]: Oops, no. I was unblock the Jenkins build. I will revert this back.

File: dev/lint-python
Comment (HyukjinKwon) [reply dev/lint-python:210]: I was about to ask Shane to install Sphinx in Python 3, and recover it back later. I will fill a JIRA and comment together.

File: python/run-tests.py
Comment (HyukjinKwon) [reply python/run-tests.py:164]: We should install `pypy3` and add it back here as well. I will also comment and file a separate JIRA.

**Comment by HyukjinKwon** [comment]
I commented and filed two JIRAs SPARK-32278 and SPARK-32279 for Jenkins env. cc @shaneknapp FYI.

Currently, PyPy 3 is not being tested in Jenkins, and PySpark documentation build via Sphinx is being skipped because Python 3 does not have Sphinx.

These will be tested via Github Actions for now via #29057.

**Comment by HyukjinKwon** [comment]
I'll merge GitHub actions one first, and fix the conflicts here

**Comment by dongjoon-hyun** [comment]
I merged GitHub Action PR. Please rebase this to the master. Thanks, @HyukjinKwon .

**Comment by HyukjinKwon** [comment]
Sure, thanks @dongjoon-hyun

File: python/pyspark/sql/tests/test_pandas_grouped_map.py
Comment (BryanCutler) [inline python/pyspark/sql/tests/test_pandas_grouped_map.py:137]: I think the default is `True`, so not needed but no big deal to leave it.

File: python/pyspark/sql/types.py
Comment (BryanCutler) [reply python/pyspark/sql/types.py:1490]: Yup, this looks good. I noticed you already fixed up the test cases that if affects, so that's great!

File: python/pyspark/sql/pandas/serializers.py
Comment (BryanCutler) [inline python/pyspark/sql/pandas/serializers.py:176]: We might want to think about removing this as an option as a followup. It was mostly added because dataframe constructed with python < 3.6 could not guarantee the order of columns, but now it should match the given schema.

**Review by BryanCutler — APPROVED** [review]
LGTM, lot's of cleanup here thanks for doing it. This is great!

File: dev/lint-python
Comment (dongjoon-hyun) [inline dev/lint-python:177]: Like this other place, shall we use the variable `PYTHON_EXECUTABLE` instead of `python3`?

File: python/pyspark/sql/pandas/serializers.py
Comment (HyukjinKwon) [reply python/pyspark/sql/pandas/serializers.py:176]: Ah, right. sounds good!

Comment (HyukjinKwon) [reply python/pyspark/sql/pandas/serializers.py:176]: and yes, maybe it should better be in a separate PR.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by BryanCutler** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
I am merging it to master. Thank you guys for reviewing this.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by holdenk** [comment]
Thanks for doing this, awesome work :)

**Comment by Fokko** [comment]
Cool stuff, thanks for the work @HyukjinKwon

**Comment by dongjoon-hyun** [comment]
Great! Thank you always for leading PySpark part (in addition to all the other Spark module), @HyukjinKwon !

File: docs/rdd-programming-guide.md
Comment (mzhang-code) [reply docs/rdd-programming-guide.md:281]: It's been long time not using pyspark, I'm glad that this is fixed in spark 3.0. Thanks.

**Pull request closed without merging**
